File: packages/sdk-go/retry_test.go
package rollgate

import (
	"context"
	"errors"
	"testing"
	"time"
)

func TestRetryConfig_Defaults(t *testing.T) {
	config := DefaultRetryConfig()

	if config.MaxRetries != 3 {
		t.Errorf("expected MaxRetries 3, got %d", config.MaxRetries)
	}
	if config.BaseDelay != 100*time.Millisecond {
		t.Errorf("expected BaseDelay 100ms, got %v", config.BaseDelay)
	}
	if config.MaxDelay != 10*time.Second {
		t.Errorf("expected MaxDelay 10s, got %v", config.MaxDelay)
	}
	if config.JitterFactor != 0.1 {
		t.Errorf("expected JitterFactor 0.1, got %f", config.JitterFactor)
	}
}

func TestCalculateBackoff(t *testing.T) {
	config := RetryConfig{
		MaxRetries:   3,
		BaseDelay:    100 * time.Millisecond,
		MaxDelay:     10 * time.Second,
		JitterFactor: 0,
	}

	t.Run("should calculate exponential backoff", func(t *testing.T) {
		// 100ms * 2^0 = 100ms
		if CalculateBackoff(0, config) != 100*time.Millisecond {
			t.Errorf("expected 100ms, got %v", CalculateBackoff(0, config))
		}
		// 100ms * 2^1 = 200ms
		if CalculateBackoff(1, config) != 200*time.Millisecond {
			t.Errorf("expected 200ms, got %v", CalculateBackoff(1, config))
		}
		// 100ms * 2^2 = 400ms
		if CalculateBackoff(2, config) != 400*time.Millisecond {
			t.Errorf("expected 400ms, got %v", CalculateBackoff(2, config))
		}
	})

	t.Run("should cap at maxDelay", func(t *testing.T) {
		shortConfig := config
		shortConfig.MaxDelay = 300 * time.Millisecond

		delay := CalculateBackoff(2, shortConfig)
		if delay != 300*time.Millisecond {
			t.Errorf("expected 300ms (capped), got %v", delay)
		}
	})

	t.Run("should add jitter", func(t *testing.T) {
		jitterConfig := config
		jitterConfig.JitterFactor = 0.5

		results := make(map[time.Duration]bool)
		for i := 0; i < 20; i++ {
			results[CalculateBackoff(0, jitterConfig)] = true
		}

		if len(results) < 2 {
			t.Error("expected jitter to produce different values")
		}
	})
}

func TestRetryer_Do(t *testing.T) {
	t.Run("should succeed on first attempt", func(t *testing.T) {
		retryer := NewRetryer(DefaultRetryConfig())
		callCount := 0

		result := retryer.Do(context.Background(), func() error {
			callCount++
			return nil
		})

		if !result.Success {
			t.Error("expected success")
		}
		if result.Attempts != 1 {
			t.Errorf("expected 1 attempt, got %d", result.Attempts)
		}
		if callCount != 1 {
			t.Errorf("expected function called once, got %d", callCount)
		}
	})

	t.Run("should retry on retryable error", func(t *testing.T) {
		config := RetryConfig{
			MaxRetries:   3,
			BaseDelay:    1 * time.Millisecond,
			MaxDelay:     10 * time.Millisecond,
			JitterFactor: 0,
		}
		retryer := NewRetryer(config)
		callCount := 0

		result := retryer.Do(context.Background(), func() error {
			callCount++
			if callCount < 3 {
				return errors.New("503 Service Unavailable")
			}
			return nil
		})

		if !result.Success {
			t.Errorf("expected success, got error: %v", result.Error)
		}
		if result.Attempts != 3 {
			t.Errorf("expected 3 attempts, got %d", result.Attempts)
		}
	})

	t.Run("should not retry on non-retryable error", func(t *testing.T) {
		retryer := NewRetryer(DefaultRetryConfig())
		callCount := 0

		result := retryer.Do(context.Background(), func() error {
			callCount++
			return errors.New("401 Unauthorized")
		})

		if result.Success {
			t.Error("expected failure")
		}
		if result.Attempts != 1 {
			t.Errorf("expected 1 attempt, got %d", result.Attempts)
		}
		if callCount != 1 {
			t.Errorf("expected function called once, got %d", callCount)
		}
	})

	t.Run("should exhaust retries", func(t *testing.T) {
		config := RetryConfig{
			MaxRetries:   2,
			BaseDelay:    1 * time.Millisecond,
			MaxDelay:     10 * time.Millisecond,
			JitterFactor: 0,
		}
		retryer := NewRetryer(config)
		callCount := 0

		result := retryer.Do(context.Background(), func() error {
			callCount++
			return errors.New("503 Service Unavailable")
		})

		if result.Success {
			t.Error("expected failure")
		}
		if result.Attempts != 3 { // 1 initial + 2 retries
			t.Errorf("expected 3 attempts, got %d", result.Attempts)
		}
		if callCount != 3 {
			t.Errorf("expected function called 3 times, got %d", callCount)
		}
	})

	t.Run("should respect context cancellation", func(t *testing.T) {
		config := RetryConfig{
			MaxRetries:   10,
			BaseDelay:    100 * time.Millisecond,
			MaxDelay:     1 * time.Second,
			JitterFactor: 0,
		}
		retryer := NewRetryer(config)

		ctx, cancel := context.WithCancel(context.Background())
		callCount := 0

		go func() {
			time.Sleep(50 * time.Millisecond)
			cancel()
		}()

		result := retryer.Do(ctx, func() error {
			callCount++
			return errors.New("503 Service Unavailable")
		})

		if result.Success {
			t.Error("expected failure due to context cancellation")
		}
		if result.Error != context.Canceled {
			t.Errorf("expected context.Canceled error, got %v", result.Error)
		}
	})
}

func TestIsRetryableError(t *testing.T) {
	tests := []struct {
		name     string
		err      error
		expected bool
	}{
		{"network timeout", errors.New("connection timeout"), true},
		{"connection refused", errors.New("connection refused"), true},
		{"503 error", errors.New("503 Service Unavailable"), true},
		{"502 error", errors.New("502 Bad Gateway"), true},
		{"504 error", errors.New("504 Gateway Timeout"), true},
		{"429 error", errors.New("429 Too Many Requests"), true},
		{"401 error", errors.New("401 Unauthorized"), false},
		{"403 error", errors.New("403 Forbidden"), false},
		{"404 error", errors.New("404 Not Found"), false},
		{"generic error", errors.New("something went wrong"), false},
		{"nil error", nil, false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := IsRetryable(tt.err)
			if result != tt.expected {
				t.Errorf("expected %v, got %v", tt.expected, result)
			}
		})
	}
}

File: packages/sdk-core/src/retry.test.ts
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import {
  RetryConfig,
  DEFAULT_RETRY_CONFIG,
  calculateBackoff,
  isRetryableError,
  fetchWithRetry,
} from './retry';

describe('retry module', () => {
  describe('DEFAULT_RETRY_CONFIG', () => {
    it('should have sensible defaults', () => {
      expect(DEFAULT_RETRY_CONFIG.maxRetries).toBe(3);
      expect(DEFAULT_RETRY_CONFIG.baseDelayMs).toBe(100);
      expect(DEFAULT_RETRY_CONFIG.maxDelayMs).toBe(10000);
      expect(DEFAULT_RETRY_CONFIG.jitterFactor).toBe(0.1);
    });
  });

  describe('calculateBackoff', () => {
    const config: RetryConfig = {
      maxRetries: 3,
      baseDelayMs: 100,
      maxDelayMs: 10000,
      jitterFactor: 0,
    };

    it('should calculate exponential backoff', () => {
      expect(calculateBackoff(0, config)).toBe(100);
      expect(calculateBackoff(1, config)).toBe(200);
      expect(calculateBackoff(2, config)).toBe(400);
      expect(calculateBackoff(3, config)).toBe(800);
    });

    it('should cap at maxDelayMs', () => {
      const shortMaxConfig: RetryConfig = {
        ...config,
        maxDelayMs: 300,
      };

      expect(calculateBackoff(2, shortMaxConfig)).toBe(300);
      expect(calculateBackoff(5, shortMaxConfig)).toBe(300);
    });

    it('should add jitter when jitterFactor > 0', () => {
      const jitterConfig: RetryConfig = {
        ...config,
        jitterFactor: 0.5,
      };

      const results = new Set<number>();
      for (let i = 0; i < 20; i++) {
        results.add(calculateBackoff(0, jitterConfig));
      }

      expect(results.size).toBeGreaterThan(1);

      for (const result of results) {
        expect(result).toBeGreaterThanOrEqual(50);
        expect(result).toBeLessThanOrEqual(150);
      }
    });

    it('should never return negative values', () => {
      const extremeJitterConfig: RetryConfig = {
        ...config,
        jitterFactor: 1,
      };

      for (let i = 0; i < 100; i++) {
        expect(calculateBackoff(0, extremeJitterConfig)).toBeGreaterThanOrEqual(0);
      }
    });
  });

  describe('isRetryableError', () => {
    it('should return true for network errors', () => {
      expect(isRetryableError(new Error('ECONNREFUSED'))).toBe(true);
      expect(isRetryableError(new Error('ETIMEDOUT'))).toBe(true);
      expect(isRetryableError(new Error('ENOTFOUND'))).toBe(true);
      expect(isRetryableError(new Error('ECONNRESET'))).toBe(true);
      expect(isRetryableError(new Error('network error'))).toBe(true);
      expect(isRetryableError(new Error('fetch failed'))).toBe(true);
    });

    it('should return true for server errors (5xx)', () => {
      expect(isRetryableError(new Error('500 Internal Server Error'))).toBe(true);
      expect(isRetryableError(new Error('502 Bad Gateway'))).toBe(true);
      expect(isRetryableError(new Error('503 Service Unavailable'))).toBe(true);
      expect(isRetryableError(new Error('504 Gateway Timeout'))).toBe(true);
    });

    it('should return true for rate limiting (429)', () => {
      expect(isRetryableError(new Error('429 Too Many Requests'))).toBe(true);
      expect(isRetryableError(new Error('too many requests'))).toBe(true);
    });

    it('should return false for client errors (4xx)', () => {
      expect(isRetryableError(new Error('400 Bad Request'))).toBe(false);
      expect(isRetryableError(new Error('401 Unauthorized'))).toBe(false);
      expect(isRetryableError(new Error('403 Forbidden'))).toBe(false);
      expect(isRetryableError(new Error('404 Not Found'))).toBe(false);
    });

    it('should return false for non-Error objects', () => {
      expect(isRetryableError('string error')).toBe(false);
      expect(isRetryableError(123)).toBe(false);
      expect(isRetryableError(null)).toBe(false);
      expect(isRetryableError(undefined)).toBe(false);
    });
  });

  describe('fetchWithRetry', () => {
    beforeEach(() => {
      vi.useFakeTimers();
    });

    afterEach(() => {
      vi.useRealTimers();
    });

    it('should succeed on first attempt', async () => {
      const fn = vi.fn().mockResolvedValue('success');

      const resultPromise = fetchWithRetry(fn);
      const result = await resultPromise;

      expect(result.success).toBe(true);
      expect(result.data).toBe('success');
      expect(result.attempts).toBe(1);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should retry on retryable error', async () => {
      const fn = vi
        .fn()
        .mockRejectedValueOnce(new Error('503 Service Unavailable'))
        .mockRejectedValueOnce(new Error('503 Service Unavailable'))
        .mockResolvedValue('success');

      const resultPromise = fetchWithRetry(fn, { maxRetries: 3, baseDelayMs: 10, jitterFactor: 0 });

      await vi.advanceTimersByTimeAsync(100);

      const result = await resultPromise;

      expect(result.success).toBe(true);
      expect(result.data).toBe('success');
      expect(result.attempts).toBe(3);
      expect(fn).toHaveBeenCalledTimes(3);
    });

    it('should not retry on non-retryable error', async () => {
      const fn = vi.fn().mockRejectedValue(new Error('401 Unauthorized'));

      const result = await fetchWithRetry(fn);

      expect(result.success).toBe(false);
      expect(result.error?.message).toBe('401 Unauthorized');
      expect(result.attempts).toBe(1);
      expect(fn).toHaveBeenCalledTimes(1);
    });

    it('should exhaust retries and fail', async () => {
      const fn = vi.fn().mockRejectedValue(new Error('503 Service Unavailable'));

      const resultPromise = fetchWithRetry(fn, { maxRetries: 2, baseDelayMs: 10, jitterFactor: 0 });

      await vi.advanceTimersByTimeAsync(1000);

      const result = await resultPromise;

      expect(result.success).toBe(false);
      expect(result.error?.message).toBe('503 Service Unavailable');
      expect(result.attempts).toBe(3);
      expect(fn).toHaveBeenCalledTimes(3);
    });
  });
});
